packages/adblocker/src/data-view.ts

```diff
@@ -389,23 +392,30 @@ export class StaticDataView {
   }
 
   public pushUTF8(raw: string): void {
-    const str = encode(raw);
-    this.pushLength(str.length);
-
-    for (let i = 0; i < str.length; i += 1) {
-      this.buffer[this.pos++] = str.charCodeAt(i);
+    const pos = this.getPos();
+    // Assume the size of output length is 1 (which means output is less than 128)
+    // based on the possible minimal length to avoid memory relocation.
+    // The minimal length is always 1 byte per character.
+    const start = pos + sizeOfLength(raw.length);
+    const { written } = TEXT_ENCODER.encodeInto(raw, this.buffer.subarray(start));
+    // If we failed to predict, that means the required bytes for length is 5.
+    if (pos + sizeOfLength(written) !== start) {
+      // Push 4 bytes back, `start + 4` or `pos + 5`
+      this.buffer.copyWithin(pos + 5, start, start + written);
     }
+    // Restore pos to push length
+    this.setPos(pos);
+    this.pushLength(written);
+    // Reflect written bytes to pos
+    this.setPos(this.pos + written);
```

Comment:
```suggestion
    this.pushLength(written);
    this.setPos(pos + written);
```

Reply:
We can't do that here. Despite `pushLength` is located later than `encodeInto`, we still want the length byte to be earlier than the actual payload. In that case, we're writing the length after the payload. Therefore, we need to recall the pos before we wrote the payload.